services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,630 @@
+    private MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Sets an the {@code MongoDbCallExecutor} and {@code MongoClient} for the mongodb connection.
+     *
+     * @param mongoDbCallExecutor An instance of the configured, connection ready mongodb executor singleton.
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = Objects.requireNonNull(mongoDbCallExecutor);
+        this.mongoClient = Objects.requireNonNull(this.mongoDbCallExecutor.getMongoClient());
+    }
+
+    public MongoDbBasedTenantsConfigProperties getConfig() {
+        return config;
+    }
+
+    /**
+     * Sets the tenant specific configuration.
+     *
+     * @param configuration The tenant configuration
+     * @throws NullPointerException if the configuration is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final MongoDbBasedTenantsConfigProperties configuration) {
+        this.config = Objects.requireNonNull(configuration);
+    }
+
+    /**
+     * Starts the service.
+     * <p>
+     * Finishes if a tenant collection is found and creates an indexed collection else.
+     * </P>
+     *
+     * @param startPromise the startup promise
+     */
+    @Override
+    public void start(final Promise<Void> startPromise) {
+        final Promise<List<String>> existingCollections = Promise.promise();
+        mongoClient.getCollections(existingCollections);
+        existingCollections.future()
+                .compose(successExistingCollections -> {
+                    if (successExistingCollections.contains(getConfig().getCollectionName())) {
+                        return Future.succeededFuture();
+                    } else {
+                        // create index & implicit collection
+                        return mongoDbCallExecutor.createCollectionIndex(getConfig().getCollectionName(),
+                                new JsonObject().put(TenantConstants.FIELD_PAYLOAD_TENANT_ID, 1).put(TenantConstants.FIELD_PAYLOAD_DEVICE_ID, 1),
+                                new IndexOptions().unique(true));
+                    }
+                })
+                .compose(success -> {
+                    if (getConfig().isModificationEnabled()) {
+                        log.info("persistence is disabled, will not save tenant identities to mongoDB.");
+                    }
+                    log.debug("startup complete");
+                    startPromise.complete();
+                    return Future.succeededFuture();
+                }).onFailure(reason -> {
+            log.error("Index creation failed", reason);
+            startPromise.fail(reason.toString());
+        });
+    }
+
+    /**
+     * Stops the service.
+     *
+     * @param stopPromise the shutdown promise
+     */
+    @Override
+    public void stop(final Promise<Void> stopPromise) {
```

Comment:
Please refer the above comment about the _javadoc_ of the `start(final Promise<Void> startPromise)`